### ISSUE TITLE
Submit training job to Google Cloud AI Platform?

### ISSUE THREAD

**Issue by robinvanschaik** [flairNLP/flair#1770]
Hi all, 

I am relatively new to FLAIR, but I think it shows great promise to me. Thanks for all the effort!

I am in the process of investigating whether it is possible to submit a FLAIR Text Classification training job to Google Cloud AI 
 Platform. I reckon that it would be nice to have the option to submit a training job in [Google Cloud AI Platform](https://cloud.google.com/blog/products/ai-machine-learning/ai-in-depth-serving-a-pytorch-text-classifier-on-ai-platform-serving-using-custom-online-prediction) in order to scale up training jobs rather than relying on Notebook VMs.

To train locally, I have my work-in-progress code wrapped up in a single function like this.

```python

def initialize_training(data_folder_path, output_folder_path, text_column_index, label_column_index,  delimiter = ';', model_type = "SentenceTransformerDocumentEmbeddings", model = 'distiluse-base-multilingual-cased', max_epochs = 10, continue_training = False):
    
    #https://cloud.google.com/ai-platform/training/docs/getting-started-scikit-xgboost
    
    '''
    Create a text classification model using FLAIR and SentenceTransformers/Huggingface Transformers.
    ------------------------
    Params: 
    data_folder_path: Folder path with each file titled appropriately i.e. train.csv test.csv dev.csv. 
                      Will create a 80/10/10 split if only train is supplied.
    output_folder_path: Folder path for storing the best model & checkpoints.
    text_column_index: In which index (starting from 0) the input column is located.
    label_column_index: In which index (starting from 0) the label column is located.
    delimiter: type of delimiter used in the .csv file.
    model_type: SentenceTransformerDocumentEmbeddings or TransformerDocumentEmbeddings
    model: Which model to use. Defaults to a multilingual model.
    
    max_epochs: Number of epochs to train the model for. 
    continue_training: Boolean. 
    ------------------------
    Output:
    best-model.pt
    final-model.pt
    
    '''
    
    # This is the folder in which train, test and dev files reside
    data_folder = data_folder_path
    
    # 1. Column format indicating which columns hold the text and label(s)
    column_name_map = {text_column_index: "text", label_column_index: "label_topic"}

    # 2. Load corpus containing training, test and dev data and if CSV has a header, you can skip it
    corpus: Corpus = CSVClassificationCorpus(data_folder,
                                             column_name_map,
                                             skip_header  = True,
                                             delimiter= delimiter # ;-separated lines
                                            ) 
    if continue_training == False:    
        # Print statistics about the corpus.
        training_data_statistics = corpus.obtain_statistics()
        print(training_data_statistics)


        # 3. Create a label dictionary.
        label_dict = corpus.make_label_dictionary()

        # 4. Initialize the sentence_transformers model.
        if model_type == "SentenceTransformerDocumentEmbeddings":
            document_embeddings = SentenceTransformerDocumentEmbeddings(model)
        elif model_type == "TransformerDocumentEmbeddings":
            document_embeddings = TransformerDocumentEmbeddings(model, fine_tune=True)
        else:
            raise Exception("Pick SentenceTransformerDocumentEmbeddings or TransformerDocumentEmbeddings.")
        
        # 5. create the text classifier
        classifier = TextClassifier(document_embeddings, label_dictionary = label_dict)

        # 6. initialize the text classifier trainer with Adam optimizer
        trainer = ModelTrainer(classifier, 
                               corpus, 
                               optimizer = Adam,
                               use_tensorboard = True)

        # 7. start the training
        trainer.train(output_folder_path,
                      learning_rate = 3e-5, # use very small learning rate
                      mini_batch_size = 16,
                      mini_batch_chunk_size = 4, # optionally set this if transformer is too much for your machine
                      max_epochs = max_epochs, # terminate after 5 epochs
                      # Save checkpoint.
                      checkpoint = True)
    
    # 8. Continue training 
    if continue_training == True:
        
        checkpoint = output_folder_path + '/checkpoint.pt'
        trainer = ModelTrainer.load_checkpoint(checkpoint, corpus)
        trainer.train(output_folder_path,
                      learning_rate = 3e-5,
                      mini_batch_size = 16,
                      mini_batch_chunk_size = 4,
                      max_epochs = max_epochs,
                      # Save checkpoint.
                      checkpoint = True)

```

However, I think I will run into a couple of issues down the line:

1. Reading .CSV stored in Google Cloud Storage as CSVClassificationCorpus requires you to pass along a folder.
2. Monitoring progress with Tensorboard requires that we dump the runs into a Cloud Storage Bucket. 
3. Storing the checkpoints in Cloud Storage Buckets. 

As far as I can tell the difficulty arises in the fact that we will need to interface with Google Cloud Storage Buckets, and I do not know how to circumvent this. 

Is there something I am overlooking that should make it possible? Or has anyone already solved this?

**Comment by alanakbik** [comment]
Hi @robinvanschaik interesting - I haven't yet worked with this but if you find solutions to these questions could you share them here? Others might find this very useful1

**Comment by robinvanschaik** [comment]
@alanakbik Thanks for the response. 
As of today I have rethought the approach that this would require and made good progress. 

It turns out that [you perform training by using docker images as input](https://cloud.google.com/ai-platform/training/docs/custom-containers-training). 

As such, I have already created a suitable docker image and set-up an MVP training script; just need to debug some things. 

I would be happy to add them to the repo, but I am not entirely sure what the guidelines are or how to proceed. 

If you have any pointers, I will get working on it for sure if I have spare time.

**Comment by alanakbik** [comment]
Thanks! I think the best would probably be a readme file and sharing the docker file in your repo. We could link to it in the Flair README (Flair landing page) so that interested persons can find the instructions and the docker file.

**Comment by robinvanschaik** [comment]
Hey @alanakbik !

I added a [repo](https://github.com/robinvanschaik/flair-on-gcp) with some files and instructions on how to submit it as a training job on GCP AI platform. 

In the meantime, I also deployed a freshly-trained Flair model as an endpoint on Cloud Run by making use of FastAPI and Docker. I will add a write-up on that as well in the repo.

**Comment by alanakbik** [comment]
Closed via #1825 - thanks @robinvanschaik!
